### PR TITLE
Error notification, TransferInfo LiveData sharing and pause, resume support

### PR DESCRIPTION
**Pull request by anuchandy**
#### Features:

1. Adding support to promote errors from Background Worker and SerialExecutor to TransferInfo LiveData
    - From Background Worker https://github.com/Azure/azure-sdk-for-android/commit/e3f433ea459b818fdb53952102dbebed53e033ea
    - From SerialExecutor https://github.com/Azure/azure-sdk-for-android/pull/224/commits/7f18255f1119be4439ef13bd552a11bc07c23087
2. Adding support for weak-reference based cache to enable sharing of TransferInfo LiveData per transfer - commit https://github.com/Azure/azure-sdk-for-android/commit/68ee38e04e403bc8a2fb3984ecc39e19c8c2d2d4
3. Adding support for pause, resume infra and enabling it for upload - commit https://github.com/Azure/azure-sdk-for-android/commit/189d1be3d051aa840ae3d5e9b228b7fb59050d6f

#### Refactor:

1. Renaming TransferIdMappedToTransferInfo to TransferIdInfoLiveData - commit https://github.com/Azure/azure-sdk-for-android/commit/d696b38f95e3b44981986e36910786ba97b6bc5f

### CODE REVIEW

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+    // flag to track whether current event about to emit from transferInfoLiveData is the first event.
+    private boolean isFirstEvent = true;
+    // flag to track whether the last emitted event from transferInfoLiveData was a pause event.
+    private boolean wasPaused;
+
+    private TransferIdInfoLiveData() {}
+
+    @MainThread
+    static TransferIdInfoLiveData.Result create(Context context) {
+        TransferIdInfoLiveData transferIdInfoLiveData = new TransferIdInfoLiveData();
```
Comment (vcolin7) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java:63]: This may be a dumb question but why are we using a static method to create this instead of making `init()` not private and using it?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+        /**
+         * Check whether the user paused the transfer in the current session.
+         * Note: this method reset the flag once called.
+         *
+         * @return true if user paused the transfer, false otherwise.
+         */
+        @MainThread
+        boolean isUserPaused() {
+            boolean b = this.userPaused;
+            this.userPaused = false;
```
Comment (vcolin7) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java:473]: Why is this reset when checking its value?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveDataCache.java [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+ * same application process. This enables all Observers of a transfer [e.g. Observers of
+ * upload(tid: 1), Observers of resume(tid: 1)] to listen to same TransferInfo LiveData,
+ * irrespective of the TransferClient they used to get the TransferInfo LiveData.
+ *
+ * The cache implementation assumes that the TransferIdOrError LiveData in the value
+ * do not strongly refer to the TransferInfo LiveData in the same value either
+ * directly or indirectly, since that will prevent the TransferIdOrError LiveData
+ * from being collected.
+ */
+final public class TransferIdInfoLiveDataCache {
```
Comment (vcolin7) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveDataCache.java:47]: What is was the reasoning behind making this class public? It seems it is only used within this package. Is the decision related to how the GC works?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferClient.java [hunk 4]
```diff
@@ -73,39 +76,133 @@ private TransferClient(Context context,
...
+
+    /**
+     * Resume a transfer that was paused.
+     *
+     * @param @param transferId the transfer id identifies the transfer to resume.
+     * @return LiveData that streams {@link TransferInfo} describing current state of the transfer
+     */
+    public LiveData<TransferInfo> resume(long transferId) {
+        // UI_Thread
+        final MutableLiveData<TransferIdOrError> idOrErrorChannel = new MutableLiveData<>();
```
Comment (vcolin7) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferClient.java:162]: Is it a convention to name MutableLiveData objects channels?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java [hunk 5]
```diff
@@ -0,0 +1,477 @@
...
+                // An error from TransferClient then emit null WorkInfo. The downstream should check error before
+                // start processing the WorkInfo.
+                return null;
+            } else {
+                final long transferId = this.inputTransferIdOrError.getId();
+                if (workInfoList == null || workInfoList.isEmpty()) {
+                    Log.e(TAG, "Received null or Empty WorkInfo list for the transfer '" + transferId + "' from WorkManager." );
+                    return null;
+                }
+                if (workInfoList.size() > 1) {
```
Comment (vcolin7) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java:274]: When would we get more than one WorkInfo? When the operation has been paused and resumed multiple times for example? If so, why do we only return the first one? Does that mean they all contain the same info?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveDataCache.java (hunk 3 above)
Comment (anuchandy) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveDataCache.java:47]: good catch! this must be a package-private, being public is a type.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferClient.java (hunk 4 above)
Comment (anuchandy) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferClient.java:162]: there is no convention like that, just my own convention, memories of Go :) We can remove that suffix.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java (hunk 5 above)
Comment (anuchandy) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java:274]: WorkManager supports chaining of workers and assigning a name to the chain, in such cases this WorkManager API returns a List of WorkInfo. In our case chaining is not used and each transfer/work has a unique name.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java (hunk 2 above)
Comment (anuchandy) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java:473]: No specific reasons just to view it as one time consumable value.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/Util.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import com.azure.android.core.internal.util.serializer.SerializerAdapter;
+import com.azure.android.core.internal.util.serializer.SerializerFormat;
+import com.azure.android.core.util.CoreUtil;
+import com.azure.android.storage.blob.models.BlobStorageException;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlRootElement;
+
+import java.util.Objects;
+
+final class Util {
```
[the remark this answers is not in the document]
Comment (anuchandy) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/Util.java:15]: My thought was to use `Util` as a place holder to add any more utility methods that we want to stay local to the package. If we come across more exception parsing scenarios then I was thinking of creating a type like `ExceptionUtil` specific for exception parsing. But I don't see any immediate use case to parse exceptions in any other format.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java (hunk 1 above)
Comment (anuchandy) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java:63]: Mainly for readability, the name `create` indicates it is factory method hence is static but `init` is initializing an already created object.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java (hunk 2 above)
Comment (anuchandy) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java:473]: and Nothing breaks if we remove it :)

Comment (anuchandy) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java:473]: now I remember why it is a one-time consumable value because if there is a `cancellation` after a `pause`, we don't want to treat that as a `pause`.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdOrError.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+ *
+ * The type used by the methods in {@link TransferClient} (e.g. upload, download, resume) to
+ * communicate data from background {@link SerialExecutor} to {@link TransferIdMappedToTransferInfo}.
+ *
+ * The data includes the operation type, the actual transfer id of a transfer or an error indicating
+ * the failure detected by TransferClient. Note that this type is used only to channel the error that
+ * TransferClient methods encountered or could detect, this type is not used to channel the real time
+ * error happened in the background worker.
+ */
+public final class TransferIdOrError {
```
Comment (bsiegel) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdOrError.java]: I feel like maybe we should just make a generic `Result<TSuccess, TErr>` type we can use instead of this. I discovered that this is somehow not a thing that Java includes itself.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java [hunk 8]
```diff
@@ -35,6 +35,9 @@
     private final MutableLiveData<TransferIdOrError> transferIdOrErrorLiveData = new MutableLiveData<>();
     // the output TransferInfo LiveData.
     private final MediatorLiveData<TransferInfo> transferInfoLiveData = new MediatorLiveData<>();
+    // the the object that TransferClient methods update any state that it want TransferInfo LiveData source
```
Comment (bsiegel) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java:48]: I don't understand this comment, can you maybe reword? It's not clear to me at this point what the TransferFlags are for. Maybe this encapsulates the different states a transfer can be in?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveDataCache.java [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+        }
+    }
+
+    /**
+     * Expunges stale entries (containing already GCed {@code TransferInfo} LiveData) from the map.
+     */
+    @SuppressWarnings("unchecked")
+    private void expunge() {
+        int cnt = 0;
+        for (Reference<?> staledReference; (staledReference = this.queue.poll()) != null;) {
```
Comment (bsiegel) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveDataCache.java:140]: Is this not just the same as a while loop? I'm fine with it but seeing this for-loop syntax is more confusing to me.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferClient.java [hunk 10]
```diff
@@ -101,16 +104,105 @@ private TransferClient(Context context,
...
-        return liveDataPair.getTransferInfoLiveData();
+        // UI_Thread
+        return toCachedTransferInfoLiveData(idOrErrorChannel, false);
     }
 
     /**
-     * Get the name for a unique transfer work.
+     * Pause a transfer identified by the given transfer id. The pause operation
+     * is a best-effort and it is possible that a transfer that is already executing
+     * may continue to transfer.
```
Comment (bsiegel) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferClient.java]: What causes the behavior described?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferClient.java (hunk 4 above)
Comment (bsiegel) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferClient.java:162]: I like the `channel` nomenclature myself. It's easier for me to think of `LiveDatas` as just channels.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java [hunk 11]
```diff
@@ -84,6 +101,15 @@ private TransferIdInfoLiveData() {}
                 this.setWasPaused(false);
                 this.transferInfoLiveData.setValue(TransferInfo.createStarted(transferId));
                 if (currentWorkInfoState == WorkInfo.State.ENQUEUED) {
+                    // the worker state can be WorkInfo.State.ENQUEUED in two cases -
```
Comment (bsiegel) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java:105]: ❤️

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java [hunk 12]
```diff
@@ -241,20 +285,44 @@ private void setLastWorkInfoState(WorkInfo.State state) {
...
+                        this.setLastWorkInfoState(WorkInfo.State.SUCCEEDED);
+                        return;
+                    }
+                }
+                this.transferInfoLiveData.setValue(TransferInfo.createFailed(this.inputTransferIdOrError.getId(),
+                    this.inputTransferIdOrError.getErrorMessage()));
+                this.setLastWorkInfoState(WorkInfo.State.FAILED);
+                return;
+            } else {
+                // TransferIdOrError.Operation.UPLOAD_DOWNLOAD
```
Comment (bsiegel) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java:312]: Unclear what this comment is for

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java (hunk 2 above)
Comment (bsiegel) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java:471]: Probably just add a clarifying comment. I'm sure this question will come up again in the future otherwise.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdOrError.java (hunk 7 above)
Comment (bsiegel) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdOrError.java]: Seeing the changes here later on. Still think we should make this a more generic name, but maybe not an actual `generic` type. Perhaps something like `TransferOperationResult`?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferClient.java (hunk 10 above)
Comment (anuchandy) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferClient.java]: it's uplifted from WorkManager `cancel` API doc. WorkManager::Worker::cancel() is best effort.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveDataCache.java (hunk 9 above)
Comment (anuchandy) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveDataCache.java:140]: I see, yea `while` is more cleaner/readable, will switch to it.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java (hunk 12 above)
Comment (anuchandy) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdInfoLiveData.java:312]: I mean, this code path hit for upload or download not for resume. I will make it more clear in the doc instead of just putting enum.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdOrError.java (hunk 7 above)
Comment (anuchandy) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/TransferIdOrError.java]: yes, I was thinking the same, generic type is something we can think if we encounter more scenarios where we need to pass value-error through LiveData. `TransferOperationResult` looks good to me.
